src/kafka-util/src/client.rs

```diff
@@ -273,30 +325,88 @@ impl<C> BrokerRewritingClientContext<C> {
     }
 }
 
-impl<C> ClientContext for BrokerRewritingClientContext<C>
+impl<C> ClientContext for TunnelingClientContext<C>
 where
     C: ClientContext,
 {
     const ENABLE_REFRESH_OAUTH_TOKEN: bool = C::ENABLE_REFRESH_OAUTH_TOKEN;
 
     fn rewrite_broker_addr(&self, addr: BrokerAddr) -> BrokerAddr {
-        match self.rewrites.get(&addr) {
-            None => addr,
-            Some(rewrite) => {
-                let rewrite = rewrite();
-                let new_addr = BrokerAddr {
-                    host: rewrite.host,
-                    port: match rewrite.port {
-                        None => addr.port.clone(),
-                        Some(port) => port.to_string(),
-                    },
-                };
-                info!(
-                    "rewriting broker {}:{} to {}:{}",
-                    addr.host, addr.port, new_addr.host, new_addr.port
-                );
-                new_addr
+        let rewrite = self.rewrites.lock().expect("poisoned").get(&addr).cloned();
+
+        let return_rewrite = |rewrite: &BrokerRewriteHandle| -> BrokerAddr {
+            let rewrite = match rewrite {
+                BrokerRewriteHandle::Simple(rewrite) => rewrite.clone(),
+                BrokerRewriteHandle::SshTunnel(ssh_tunnel) => {
+                    // The port for this can change over time, as the ssh tunnel is maintained through
+                    // errors.
+                    let addr = ssh_tunnel.local_addr();
+                    BrokerRewrite {
+                        host: addr.ip().to_string(),
+                        port: Some(addr.port()),
+                    }
+                }
+            };
+
+            let new_addr = BrokerAddr {
+                host: rewrite.host.clone(),
```

Comment:
Just noticed: you already clone the `rewrite` above—can you not just move `rewrite.host` here?

Reply:
might be able to, ill check!